pkg/exchange/okex/stream_test.go

```diff
@@ -0,0 +1,552 @@
+package okex
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/c9s/bbgo/pkg/testutil"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+func getTestClientOrSkip(t *testing.T) *Stream {
+	if b, _ := strconv.ParseBool(os.Getenv("CI")); b {
+		t.Skip("skip test for CI")
+	}
+
+	key, secret, passphrase, ok := testutil.IntegrationTestWithPassphraseConfigured(t, "OKEX")
+	if !ok {
+		t.Skip("Please configure all credentials about OKEX")
+		return nil
+	}
+
+	return NewStream(key, secret, passphrase)
+}
+
+func TestStream(t *testing.T) {
+	// t.Skip()
+	s := getTestClientOrSkip(t)
+
+	// symbols := []string{
+	// 	"BTCUSDT",
+	// 	"ETHUSDT",
+	// 	"DOTUSDT",
+	// 	"ADAUSDT",
+	// 	"AAVEUSDT",
+	// 	"APTUSDT",
+	// 	"ATOMUSDT",
+	// 	"AXSUSDT",
+	// 	"BNBUSDT",
+	// 	"SOLUSDT",
+	// 	"DOGEUSDT",
+	// }
+
+	t.Run("Auth test", func(t *testing.T) {
+		s.Connect(context.Background())
+		c := make(chan struct{})
+		<-c
+	})
+
+	t.Run("book test", func(t *testing.T) {
+		s.Subscribe(types.BookChannel, "BTCUSDT", types.SubscribeOptions{
+			Depth: types.DepthLevel50,
+		})
+		s.SetPublicOnly()
+		err := s.Connect(context.Background())
+		assert.NoError(t, err)
+
+		s.OnBookSnapshot(func(book types.SliceOrderBook) {
+			t.Log("got snapshot", book)
+		})
+		s.OnBookUpdate(func(book types.SliceOrderBook) {
+			t.Log("got update", book)
+		})
+		c := make(chan struct{})
+		<-c
+	})
+
+	// t.Run("book test on unsubscribe and reconnect", func(t *testing.T) {
+	// 	for _, symbol := range symbols {
+	// 		s.Subscribe(types.BookChannel, symbol, types.SubscribeOptions{
+	// 			Depth: types.DepthLevel50,
+	// 		})
+	// 	}
+
+	// 	s.SetPublicOnly()
+	// 	err := s.Connect(context.Background())
+	// 	assert.NoError(t, err)
+
+	// 	s.OnBookSnapshot(func(book types.SliceOrderBook) {
+	// 		t.Log("got snapshot", book)
+	// 	})
+	// 	s.OnBookUpdate(func(book types.SliceOrderBook) {
+	// 		t.Log("got update", book)
+	// 	})
+
+	// 	<-time.After(2 * time.Second)
+
+	// 	s.Unsubscribe()
+	// 	for _, symbol := range symbols {
+	// 		s.Subscribe(types.BookChannel, symbol, types.SubscribeOptions{
+	// 			Depth: types.DepthLevel50,
+	// 		})
+	// 	}
+
+	// 	<-time.After(2 * time.Second)
+
+	// 	s.Reconnect()
+
+	// 	c := make(chan struct{})
+	// 	<-c
+	// })
+
+	t.Run("market trade test", func(t *testing.T) {
+		s.Subscribe(types.MarketTradeChannel, "BTCUSDT", types.SubscribeOptions{})
+		s.SetPublicOnly()
+		err := s.Connect(context.Background())
+		assert.NoError(t, err)
+
+		s.OnMarketTrade(func(trade types.Trade) {
+			t.Log("got update", trade)
+		})
+		c := make(chan struct{})
+		<-c
+	})
+
+	t.Run("wallet test", func(t *testing.T) {
+		s.OnAuth(func() {
+			t.Log("authenticated")
+		})
+		err := s.Connect(context.Background())
+		assert.NoError(t, err)
+
+		s.OnBalanceSnapshot(func(balances types.BalanceMap) {
```

Comment:
Does it support BalanceUpdate?

Reply:
Balance and position channel would update, but i'm not yet to implement ( https://www.okx.com/docs-v5/en/?shell#trading-account-websocket-positions-channel )